app/code/Magento/MediaGallery/Model/ResourceModel/SearchAssets.php

```diff
@@ -0,0 +1,173 @@
+use Magento\Framework\Api\SearchCriteriaInterface;
+use Magento\Framework\Api\Search\SearchResultFactory;
+use Magento\Framework\DB\Select;
+use Magento\MediaGalleryApi\Api\Data\AssetInterface;
+
+/**
+ * Get media assets by searchCriteria
+ */
+class SearchAssets implements SearchAssetsInterface
+{
+    private const TABLE_MEDIA_GALLERY_ASSET = 'media_gallery_asset';
+
+    /**
+     * @var ResourceConnection
+     */
+    private $resourceConnection;
+
+    /**
+     * @var AssetInterfaceFactory
+     */
+    private $mediaAssetFactory;
+
+    /**
+     * @var SearchResultFactory
+     */
+    protected $searchResultFactory;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * @param SearchResultFactory $searchResultFactory
+     * @param ResourceConnection $resourceConnection
+     * @param AssetInterfaceFactory $mediaAssetFactory
+     * @param LoggerInterface $logger
+     */
+    public function __construct(
+        SearchResultFactory $searchResultFactory,
+        ResourceConnection $resourceConnection,
+        AssetInterfaceFactory $mediaAssetFactory,
+        LoggerInterface $logger
+    ) {
+        $this->searchResultFactory = $searchResultFactory;
+        $this->resourceConnection = $resourceConnection;
+        $this->mediaAssetFactory = $mediaAssetFactory;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function execute(SearchCriteriaInterface $searchCriteria): array
+    {
+        $assets = [];
+        try {
+            foreach ($this->getAssetsData($searchCriteria)->getItems() as $assetData) {
+                $assets[] = $this->mediaAssetFactory->create(
+                    [
+                        'id' => $assetData['id'],
+                        'path' => $assetData['path'],
+                        'title' => $assetData['title'],
+                        'description' => $assetData['description'],
+                        'source' => $assetData['source'],
+                        'hash' => $assetData['hash'],
+                        'contentType' => $assetData['content_type'],
+                        'width' => $assetData['width'],
+                        'height' => $assetData['height'],
+                        'size' => $assetData['size'],
+                        'createdAt' => $assetData['created_at'],
+                        'updatedAt' => $assetData['updated_at'],
+                    ]
+                );
+            }
+        } catch (\Exception $exception) {
+            $this->logger->critical($exception);
+            throw new LocalizedException(__('Could not retrieve media assets'), $exception->getMessage());
+        }
+        return $assets;
+    }
+
+    /**
+     * Retrieve assets data from database
+     *
+     * @param SearchCriteriaInterface $searchCriteria
+     * @return SearchResultInterface
+     */
+    private function getAssetsData(SearchCriteriaInterface $searchCriteria): SearchResultInterface
```

Comment:
please extract SeachCriteria processing to a separate ResourceModel class (while the main service implementation class can be in the Model namespace)